x-pack/plugins/ml/server/lib/alerts/jobs_health_service.ts

```diff
@@ -154,12 +159,59 @@ export function jobsHealthServiceProvider(
           };
         });
     },
+    /**
+     * Returns annotations about delayed data
+     * @param jobIds
+     * @param timeInterval
+     * @param previousStartedAt
+     */
+    async getDelayedDataReport(
+      jobIds: string[],
+      timeInterval: string | null,
+      previousStartedAt: Date | null,
+      docsCount: number | null
+    ): Promise<DelayedDataResponse[]> {
+      const currentTimestamp = Date.now();
+
+      let earliestMs;
+
+      if (timeInterval) {
+        earliestMs = currentTimestamp - parseInterval(timeInterval)!.asMilliseconds();
+      } else if (previousStartedAt) {
+        earliestMs = previousStartedAt.getTime();
+      }
```

Comment:
My assumptions:
 - timeInterval is the window of time the user cares about delayed data in the past
 - previousStartedAt is the time of the LAST firing of this alert check

Seems to me a couple of things:

1. `timeInterval` should never be null. 
2. The query should check the following (and it doesn't)

`endTime >= parseInterval(timeInterval)!.asMilliseconds()`

`modifiedTimeStamp >= previousStartedAt.getTime()`

`endTime` should never be compared against the "previous fire timestamp" of the alert.

You should end up with two timestamps to compare:

- Has the document been updated since our last check? (modified time vs alert check time)
- Is the latest bucket within our window of concern? (end time vs time interval)

Reply:
> timeInterval is the window of time the user cares about delayed data in the past
correct 
> previousStartedAt is the time of the LAST firing of this alert check
no, it's the date object of the previous check, not necessarily triggered the alert instance. 

> timeInterval should never be null.
it can be null, in that case, it's derived from the maximum bucket span. 

We also need to take into account the alerting rule interval check. So I think it'd be easier just to resolve the lookback interval based on the bucket span length and make in case the user has provided a shorter time interval, fallback to the double bucket span. A longer custom interval is fine.